### PR TITLE
llama : refactor sampling_info to use buffer_view template

### PR DESCRIPTION
**Pull request by danbev**
This commit updates the sampling_info struct in llama-context to use a buffer_view template for the logits, probs, sampled tokens, and candidates buffers.

The motivation for this is to simplify the code, improve type safety and readability.

### CODE REVIEW

File: src/llama-context.h [hunk 1]
```diff
@@ -238,6 +238,16 @@ struct llama_context {
...
+    template <typename T>
+    struct buffer_view {
+        T * data;
+        size_t size = 0;
+
+        bool has_data() const {
+            return data && size > 0;
+        }
+    };
+
```
Comment (ggerganov) [inline src/llama-context.h]: Better to be in `llama-impl.h`

Comment (danbev) [reply src/llama-context.h]: @ggerganov I've moved this into llama-impl.h now. Please take a another look when you get a chance and let me know if there is anything else that needs changing.

**Review by ggerganov — COMMENTED** [review]
For consistency, the non-sampling logits and embeddings could be wrapped in buffer view as well

**Review by ggerganov — APPROVED** [review]
Can merge if the CI is green
